delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storages/storage/Storages.scala

```diff
@@ -0,0 +1,632 @@
+    * @param projects    a projects operations bundle
+    * @param access      a function to verify the access to a certain storage
+    */
+  final def apply(
+      config: StoragesConfig,
+      eventLog: EventLog[Envelope[StorageEvent]],
+      permissions: Permissions,
+      orgs: Organizations,
+      projects: Projects,
+      access: StorageAccess = StorageAccess.apply
+  )(implicit
+      uuidF: UUIDF,
+      clock: Clock[UIO],
+      scheduler: Scheduler,
+      as: ActorSystem[Nothing],
+      rcr: RemoteContextResolution
+  ): UIO[Storages] =
+    for {
+      agg     <- aggregate(config, access, permissions)
+      index    = cache(config)
+      storages = apply(agg, eventLog, index, orgs, projects)
+      _       <- UIO.delay(startIndexing(config, eventLog, index, storages))
+    } yield storages
+
+  private def apply(
+      agg: StoragesAggregate,
+      eventLog: EventLog[Envelope[StorageEvent]],
+      index: StoragesCache,
+      orgs: Organizations,
+      projects: Projects
+  )(implicit rcr: RemoteContextResolution, uuidF: UUIDF) =
+    new Storages(agg, eventLog, index, orgs, projects)
+
+  private def cache(config: StoragesConfig)(implicit as: ActorSystem[Nothing]): StoragesCache = {
+    implicit val cfg: KeyValueStoreConfig      = config.keyValueStore
+    val clock: (Long, StorageResource) => Long = (_, resource) => resource.rev
+    KeyValueStore.distributed(moduleType, clock)
+  }
+
+  private def startIndexing(
+      config: StoragesConfig,
+      eventLog: EventLog[Envelope[StorageEvent]],
+      index: StoragesCache,
+      storages: Storages
+  )(implicit as: ActorSystem[Nothing], sc: Scheduler) =
+    StreamSupervisor.runAsSingleton(
+      "StorageIndex",
+      streamTask = Task.delay(
+        eventLog
+          .eventsByTag(moduleType, Offset.noOffset)
+          .mapAsync(config.indexing.concurrency)(envelope =>
+            storages
+              .fetch(IriSegment(envelope.event.id), envelope.event.project)
+              .redeemCauseWith(_ => IO.unit, res => index.put(StorageKey(res.value.project, res.value.id), res))
+          )
+      ),
+      retryStrategy = RetryStrategy(
+        config.indexing.retry,
+        _ => true,
+        RetryStrategy.logError(logger, "storages indexing")
+      )
+    )
+
+  private def aggregate(config: StoragesConfig, access: StorageAccess, permissions: Permissions)(implicit
+      as: ActorSystem[Nothing],
+      clock: Clock[UIO]
+  ) = {
+    val definition = PersistentEventDefinition(
+      entityType = moduleType,
+      initialState = Initial,
+      next = next,
+      evaluate = evaluate(access, permissions, config.storageTypeConfig),
+      tagger = (event: StorageEvent) =>
+        Set(
+          moduleType,
+          s"${Projects.moduleType}=${event.project}",
+          s"${Organizations.moduleType}=${event.project.organization}"
+        ),
+      snapshotStrategy = config.aggregate.snapshotStrategy.combinedStrategy(
```

Comment:
I don't think there's a need to snapshot storages; the expected number of events is quite low.